frontend/catalyst/compilation_pipelines.py

```diff
@@ -688,13 +725,33 @@ def get_cmain(self, *args):
         return function.get_cmain(*args)
 
     def __call__(self, *args, **kwargs):
+        static_args_hash = tuple()
+        if self.compile_options.static_argnums:
+            # Build hash for multiple static arguments.
+            static_argnums = self.compile_options.static_argnums
+            static_args_hash = tuple(hash(args[idx]) for idx in range(len(args))\
```

Comment:
What I don't understand yet is how the hash affects the decision to recompile or not. I think that decision is made in `_ensure_real_arguments_and_formal_parameters_are_compatible` but I don't see the hash values being taken into account there 🤔

Reply:
There is a hash table checking right after this part. If nothing is found in the hash table ``self.compiled_function`` will be set to ``None``.  ``self.compiled_function`` is passed to ``_ensure_real_arguments_and_formal_parameters_are_compatible`` and will trigger recompilation. I think I should integrate the has into ``_ensure_real_arguments_and_formal_parameters_are_compatible``.

Reply:
Ah you're absolutely right, thanks for pointing that out!

Could you add a comment to this branch specifying that it will automatically trigger recompilation in the `_ensure_real_arguments_and_formal_parameters_are_compatible` function?
https://github.com/PennyLaneAI/catalyst/blob/5de45605eee27589e3b031f4f6abdd59afb75118/frontend/catalyst/compilation_pipelines.py#L738-L741

Reply:
Oh never mind, nice refactor 💯